File: NeetCode/Greedy/Hand_Of_Straights/Hand_Of_Straights.cpp
// Hand_Of_Straights.cpp
//
// Compile locally with:
// clang++ Hand_Of_Straights.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    bool isNStraightHand(vector<int> &hand, int groupSize)
    {
        int n = hand.size();
        if (n % groupSize != 0)
        {
            return false;
        }

        unordered_map<int, int> count;

        for (int card : hand)
        {
            count[card]++;
        }

        vector<int> keys;

        for (const auto &[card, count] : count)
        {
            keys.push_back(card);
        }
        sort(keys.begin(), keys.end());

        for (int cardVal : keys)
        {
            while (count[cardVal] > 0)
            {
                for (int x = cardVal; x < cardVal + groupSize; x++)
                {
                    if (count[x] == 0)
                    {
                        return false;
                    }
                    count[x]--;
                }
            }
        }
        return true;
    };

    int main()
    {
        cout << "Running Hand_Of_Straights.cpp..." << endl;

        Solution sol;

        // TODO:
        // Add local test calls here

        return 0;
    }

File: NeetCode/Greedy/Hand_Of_Straights/hand__of__straights.py
from typing import Optional, List
import heapq
from collections import defaultdict, Counter


class Solution:
    def isNStraightHand(self, hand: List[int], groupSize: int) -> bool:
        if len(hand) % groupSize != 0:
            return False

        count = Counter(hand)

        for card in sorted(count):
            while count[card] > 0:
                for x in range(card, card + groupSize):
                    if count[x] == 0:
                        return False
                    count[x] -= 1
        return True


if __name__ == "__main__":
    sol = Solution()
    print("Running Solution...")


"""_summary_

class Solution:
    def isNStraightHand(self, hand: List[int], groupSize: int) -> bool:
        hand.sort()

        currSize = 1
        last = hand.pop(0)
        ptr = 0

        while hand and ptr < len(hand):
            if currSize == groupSize:
                currSize = 0
                ptr = 0
                last = -1
                continue
            if hand[ptr] == last + 1 or last == -1:
                last = hand.pop(ptr)
                currSize += 1
                continue
            ptr += 1

        return currSize == groupSize and len(hand) == 0

        """
